mesop/runtime/context.py

```diff
@@ -37,15 +37,65 @@ def __init__(
     self._node_slot_children_count: int | None = None
     self._viewport_size: pb.ViewportSize | None = None
     self._theme_settings: pb.ThemeSettings | None = None
+    self._query_params: dict[str, list[str]] = {}
+
+  def query_params(self) -> dict[str, list[str]]:
+    return self._query_params
+
+  def initialize_query_params(self, query_params: Sequence[pb.QueryParam]):
+    for query_param in query_params:
+      self._query_params[query_param.key] = list(query_param.values)
+
+  def set_query_param(self, key: str, value: str | list[str] | None):
+    if value is None:
+      del self._query_params[key]
+    else:
+      self._query_params[key] = [value] if isinstance(value, str) else value
+    self._commands.append(
+      pb.Command(
+        update_query_param=pb.UpdateQueryParam(
+          query_param=pb.QueryParam(
+            key=key, values=[value] if isinstance(value, str) else value
+          )
+        )
+      )
+    )
 
   def commands(self) -> list[pb.Command]:
     return self._commands
 
   def clear_commands(self) -> None:
     self._commands = []
 
-  def navigate(self, url: str) -> None:
-    self._commands.append(pb.Command(navigate=pb.NavigateCommand(url=url)))
+  def navigate(
+    self, url: str, query_params: dict[str, str | list[str]] | None = None
+  ) -> None:
+    query_param_protos = None
+    if query_params is not None:
+      query_param_protos = [
+        pb.QueryParam(
+          key=key, values=[value] if isinstance(value, str) else value
+        )
+        for key, value in query_params.items()
+      ]
+    # Construct the full URL with query parameters
+    full_url = url
+    if query_params:
+      query_string = "&".join(
+        f"{key}={value}"
+        if isinstance(value, str)
+        else "&".join(f"{key}={v}" for v in value)
```

Comment:
I think key/value may need to be url encoded and the key may need to be possible validated (is there a specific format the key must follow?) 

There is one case where a user may use "&" as a value.

Reply:
Good callout - I added a test case, turns out it's handled by URLSearchParams ([spec](https://url.spec.whatwg.org/#urlsearchparams-stringification-behavior)) on the TS side. This is nice because the Python side doesn't need to worry about URL encoding and decoding and we use the browser APIs.